File: march-2026/3-18-flight-vouchers/flight-vouchers.py
# Flight Vouchers 🏖️
# Alan

def pick_voucher(vouchers, delays, max_wait):
  best_option_index = -1
  best_option_value = -1

  for i in range(len(vouchers)):
    if delays[i] <= max_wait and delays[i] > 0:
      dollars_per_hour_waited = vouchers[i] / delays[i]
      if dollars_per_hour_waited > best_option_value:
        best_option_index = i
        best_option_value = dollars_per_hour_waited

  return best_option_index

File: march-2026/3-18-flight-vouchers/flight_vouchers.js
// Flight Vouchers 🏖️ 
// shercodes

function pickVoucher(vouchers, delays, max_wait) {
  // Sets the values for bestIndex and maxRatio
  let bestIndex = -1;   // If none options qualify
  let maxRatio = -1;    // The highest ratio must be greater than -1
  
  // Filters out any delay time that exceeds max_wait
  for (let i = 0; i < delays.length; i++) {
    if (delays[i] <= max_wait) { 
        
      // Calculates the ratio of each valid voucher
      const ratio = vouchers[i] / delays[i];

      // Finds the voucher with the highest ratio
      if (ratio > maxRatio) {
        maxRatio = ratio;
        bestIndex = i;
      }
    }
  }
  
  return bestIndex;
};
